src/middleware/cache/readThroughCacheMiddleware.ts

```diff
@@ -5,55 +5,76 @@ import {
 } from "http-proxy-middleware";
 import { Request, Response } from "express";
 import { featuresCache } from "../../services/cache";
-import { registrar } from "../../services/registrar";
 import logger from "../../services/logger";
 import { eventStreamManager } from "../../services/eventStreamManager";
 
 const scopedMiddlewares: Record<string, RequestHandler> = {};
+const errorCounts: Record<string, number> = {};
 
-const interceptor = responseInterceptor(
-  async (responseBuffer, proxyRes, req: Request, res: Response) => {
-    const response = responseBuffer.toString("utf-8");
-    if (!featuresCache) {
-      return response;
-    }
+const interceptor = (proxyTarget: string) =>
+  responseInterceptor(
+    async (responseBuffer, proxyRes, req: Request, res: Response) => {
+      // got response, reset error count
+      errorCounts[proxyTarget] = 0;
 
-    if (res.statusCode === 200) {
-      logger.debug("cache MISS, setting cache...");
-
-      // refresh the cache
-      try {
-        const apiKey = res.locals.apiKey;
-        const responseJson = JSON.parse(response);
-        const oldEntry = await featuresCache.get(apiKey);
-
-        featuresCache
-          .set(apiKey, responseJson)
-          .catch((e) => logger.error(e, "Unable to set cache"));
-
-        eventStreamManager.publish(
-          apiKey,
-          "features",
-          responseJson,
-          oldEntry?.payload
-        );
-      } catch (e) {
-        logger.error(e, "Unable to parse response");
+      const response = responseBuffer.toString("utf-8");
+      if (!featuresCache) {
+        return response;
       }
+
+      if (res.statusCode === 200) {
+        logger.debug("cache MISS, setting cache...");
+
+        // refresh the cache
+        try {
+          const apiKey = res.locals.apiKey;
+          const responseJson = JSON.parse(response);
+          const oldEntry = await featuresCache.get(apiKey);
+
+          featuresCache
+            .set(apiKey, responseJson)
+            .catch((e) => logger.error(e, "Unable to set cache"));
+
+          eventStreamManager.publish(
+            apiKey,
+            "features",
+            responseJson,
+            oldEntry?.payload
+          );
+        } catch (e) {
+          logger.error(e, "Unable to parse response");
+        }
+      }
+      return response;
     }
-    return response;
+  );
+
+export default async ({ proxyTarget }: { proxyTarget: string }) => {
+  // slow down requests to GrowthBook API if error counts are too high
+  const errorCount = errorCounts[proxyTarget] || 0;
+  if (errorCount > 5) {
+    const delaySecs = Math.pow(
+      1.5,
+      Math.min(errorCount, 8) + Math.random() - 5
```

Comment:
This doesn't produce very much randomness.  After 8 errors, the delay ranges from 3.8 to 5 seconds.  In a cache stampede scenario, all the requests coming in at once as opposed to spread out over 1.2 seconds doesn't seem like a big difference.

Reply:
I wasn’t really sure what delay/random profile we were looking for. Are you thinking like 10s +/- 5s when maxed out?